### ISSUE TITLE
Allow some ChargeIncrementHandler attributes to be automatically set

### ISSUE THREAD

**Issue by mattwthompson** [openforcefield/openforcefield#721]
**Is your feature request related to a problem? Please describe.**
@SimonBoothroyd says:
> Would you consider extending the charge increment handler so as to allow certain attributes to be automatically set? Namely so I can just set `charge_increment1` to the desired value and specify that `charge_increment2` should automatically be chosen to be negative that value without have to write `charge_increment2=-X` in the file?
This would simplify optimisations a bit (currently a FB eval attribute needs to be used) and for sure make computing derivatives w.r.t. these parameters easier
Otherwise this behaviour needs to be hardcoded into places like evaluator, and I'm not sure have well the current behaviour plays with autodiffing

**Describe the solution you'd like**
When N atoms are tagged, but N-1 charge increments are specified, the atom without a specified charge increments gets a charge determined by adding up all other charge increments and multiplying by -1. So, a parameter like 

```XML
<ChargeIncrement smirks="[#6X4:1]-[#8:2]" charge1increment="-0.0718" charge2increment="+0.0718"/>`
```
could be equivalently be expressed as
```XML
<ChargeIncrement smirks="[#6X4:1]-[#8:2]" charge1increment="-0.0718"/>
```
while only exposing one value to optimization as `charge2increment` is automatically set. If N-2 are specified, raise an exception (under discussion, see below). 

**Describe alternatives you've considered**
* An approach like above but without the N-1 restriction should be considered. @j-wags summarizes this proposed implementation: 

> Hmm, so Simon's proposing, for a `ChargeIncrementType` with N tagged atoms, allowing the specification of between 1 and N `chargeincrementX` values. Then every atom with an explcitly-specified `chargeincrementX` value is guaranteed to get exactly the specified increment, and all the others get an equal share of the offset required to make the whole thing net-neutral?

which might be useful for ACC fitting. A characteristic of this approach is that the remaining charge is spread out **evenly** among the other atoms, which should be made clear to the user in documentation, and would also not error out in the N-2 case mentioned above. 

* A more elegant and general approach would be to allow defining variables and arithmetic operations inside the `ChargeIncrementType` along the lines of `<ChargeIncrement smirks="[#6X4:1]-[#6X3a:2]" n="0.01" charge1increment="-1*n" charge2increment="+1*n"/>`. This would enable some fun ideas to be implemented, but also required adding a general expression parser. This may even be overkill for fitting until somebody wishes to do something like charge corrections with highly complex rules.

**Additional context**
Constraints here are
1. It would be nice (probably necessary) for OFFXMLs to survive round-trips (OFFXML -> `ForceField` -> OFFXML) for working well with ForceBalance and other libraries
2. There needs to be a way to carry this information into the System object for differentiation purposes. This should be doable with a constraint matrix
3. This change in behavior likely impacts parsing truly mangled OFFXMLs, so we'd want to consider cases in which an invalid parameter with the current spec would be read as valid after this change.

**Comment by davidlmobley** [comment]
I like this idea in general. I can see how it would greatly simplify simple fitting cases, and it doesn't add a lot of additional complexity like some of the alternatives.

**Comment by mrshirts** [comment]
> When N atoms are tagged, but N-1 charge increments are specified, the atom without a specified charge increments gets a charge determined by adding up all other charge increments and multiplying by -1

Do we need this?  I'd rather have a specification that just gives an error if they don't all add up together, and have some other function that is in charge of setting them.  I'm confused by why this situation would ever occur, where N atoms are tagged, but N-1.

Or are there instances where N=/=2?   If it's just a matter of specification in OpenFF, if N==2 always, then why not something like:

`<ChargeIncrement smirks="[#6X4:1]-[#8:2]" chargeincrement="0.0718"/>`
Which automatically always does a +1*(chargeincrement) to the first atom and -1*(chargeincrement) to the second atom?

Are there issues where a smirks expression can be in either order?

I think that 3 atom charge increment can always be decomposed into 2  sets of 2 atom charge increments?

**Comment by mattwthompson** [comment]
@SimonBoothroyd will provide more coherent answers but I'll try explaining to the best of my understanding
 
> Do we need this?

I think the current use case is when ForceBalance is trying to optimize a BCC-like parameter, in which the two values are naturally coupled but exposed as separate parameters in optimization, introducing an unnecessary degree of freedom that needs to be clamped down later by the optimizer and/or produces issues with differentiation down the pipeline. Not sure if this amounts to a need

> Or are there instances where N=/=2?

In practice it may only ever be 2 now, but the [spec](https://github.com/openforcefield/openforcefield/blob/07ee344344c9f9364569b07e7dcfd157570a8fa9/The-SMIRNOFF-force-field-format.md#chargeincrementmodel-small-molecule-and-fragment-charges) doesn't include that restriction and it may be useful in future for fitting ACCs or biopolymers. There may be situations in which N is large (an entire residue?) and cumbersome (but not impossible) to decompose into a series of 2-atom increments.

**Comment by ocmadin** [comment]
> think the current use case is when ForceBalance is trying to optimize a BCC-like parameter, in which the two values are naturally coupled but exposed as separate parameters in optimization, introducing an unnecessary degree of freedom that needs to be clamped down later by the optimizer and/or produces issues with differentiation down the pipeline.

To my understanding, this is the reason we would need this.  Having coupled parameters would require fixing many of them beforehand in ForceBalance, which would be cumbersome.

**Comment by davidlmobley** [comment]
I agree with Owen and Matt: I think it will help ForceBalance because, if you're only dealing with a single BCC like parameter involving only two atoms, having it represented in the XML as two parameters means there are two things FB has to change in tandem, which will likely not be supported at present. OTOH if it's just a single parameter (as it actually is in such cases) means FB presently can do it.

Also, Cerutti just mentioned a case he's thinking of where N would NOT be 2. So we want to leave this possibility open.

**Comment by SimonBoothroyd** [comment]
> FB has to change in tandem, which will likely not be supported at present. OTOH if it's just a single parameter (as it actually is in such cases) means FB presently can do it.

Just to note FB will support it (you can use an `eval` statement to set one parameter equal to the negative of the other), however `evaluator` doesn't currently when computing derivatives as it doesn't 'know' that `charge_increment1` and `charge_increment2` (and more if N > 2) are coupled (i.e. they _could_ in principle be decoupled, not that I'm sure they ever would be). In principle this behaviour could be hardcoded into evaluator however I think the solution proposed here would be preferable so that `evaluator` isn't having to make assumptions about any force field parameters / handlers. The system object would likely run into a similar issue unless separate information like a constraint matrix is provided.

**Comment by j-wags** [comment]
> Or are there instances where N=/=2?

As others have said, the current implementation allows for N tagged atoms in a single chargeincrement. All other things equal, I'd like to keep the implementation of this new feature general and applicable to groups on N atoms.

**Comment by mrshirts** [comment]
So, that's a good question.  Is a set of arbitrary charge increments equivalent to bond increments?  I think that it can be rewritten in terms of pairwise increments.  If we have N charges that sum to 0, then we have N-1 variables.  If there are at least N-1 bonds, then we can specify all of them.  I think that a molecule with N atoms must have at minimum N-1 bonds? or else it can't be connected?  Maybe I'm missing something important fact here. 

Example:  molecule A-B-C has "base" charges zero, then BCC's adds -0.4 to A and +0.3 to C.  By neutrality, B must have +0.1.  But if we have BCC's A-B=-0.4 and B-C= -0.3, we get the same charge +0.1 on B (convention is you get + charge if in front, -1 charge if 2nd). 

4 atoms: A-B-C-D, with charges  -0.4, -0.8, +0.5, +0.7.  Then A-B would have BCC -0.4, C-D would be -0.7,   
So we have one BCC left over, B-C=x, with the requirement that +0.4 + x = -0.8 and -x-0.7=+0.5.  These are both satisfied by x -1.2.                                 
                                                                                                                                                                     B
                                                                                                                                                                   /     \
The interesting case is more than N-1 bonds.  Let's take the first case, except the topology is  A----C, so we have N atoms and N bonds.   

Then with those same three charges we can use those bond corrections A-B=-0.4 and B-C= +0.3, and C-A=0, but this ALSO works if C-A=+0.1, B-C = -0.2, and A-B =-0.3.  So there is a degeneracy of description, with multiple BCC's describing the same situation which is bad for optimization.

So I guess the conclusion is that we may NEED to specify atoms in sets of more than 2, or the optimization will fail because of degeneracies. 

Seems like Chris would have thought about this?

So I guess its best if we do allow the remaining charges to be filled in automatically.

**Comment by j-wags** [comment]
>  I think that a molecule with N atoms must have at minimum N-1 bonds?
>  So there is a degeneracy of description, with multiple BCC's describing the same situation which is bad for optimization.
> So I guess the conclusion is that we may NEED to specify atoms in sets of more than 2, or the optimization will fail because of degeneracies.

I pretty much agree, though important to note that I think that the possibility of degenerate representations of charge transfer is inevitable. Chargeincrements are "special" compared to other SMIRNOFF terms in that they allow "partial overlaps". So molecule `A-B-C` can be simultaneously assigned lots of possible chargeincrements -- one on `A-B`, another on `B-C` another on `A-C`, and another on `A-B-C`. Unlike bonds, angles, and proper torsions, where the Topology tells us _exactly_ how many parameters we expect to be in the final System, an unknown number of chargeincrements can be applied to a given molecule. As a consequence, it's not clear to me that it's possible to avoid "triangle-degeneracy" cases (or the equivalent problem with 4+ edges/nodes), except by being careful about the SMIRKS that a user seeds their OFFXML with.

> So I guess its best if we do allow the remaining charges to be filled in automatically.

Agree. On an implementation level, supporting the `N-1` case could be implemented quickly and reverse-compatibly. Allowing for more than one "missing" chargeincrement will require a "how to fill in the rest" scheme, which would add a new keyword somewhere, which we'd need to discuss. So, as a quick fix, I'd be in favor of the N-1. I'll get started implementing that now :-)

**Comment by j-wags** [comment]
@SimonBoothroyd Due to how our IndexedParameterAttribute class works, the defined `chargeincrements` will have to be contiguously-indexed, starting at 1. That means that the highest-indexed tagged atom will be the one getting assigned the "offsetting" charge increment. Is that OK for your use case?

**Comment by mattwthompson** [comment]
I believe this was implemented in #726. Feel free to re-open if there is something outstanding or not yet implemented
